packages/core/src/browser/widget-manager.ts

```diff
@@ -222,6 +218,34 @@ export class WidgetManager {
         }
     }
 
+    /**
+     * Creates a new widget or returns the existing widget for the given description.
+     * @param factoryId the widget factory id.
+     * @param options the widget factory specific information.
+     *
+     * @returns a promise resolving to the widget.
+     */
+    async getOrCreateWidget<T extends Widget>(factoryId: string, options?: any): Promise<T> {
+        const key = this.toKey({ factoryId, options });
+        const existingWidget = this.doGetWidget<T>(key);
+        if (existingWidget) {
+            return existingWidget;
+        }
+        return this.createWidget(factoryId, options);
+    }
+
+    /**
+     * Returns whether the widget exists or not.
+     * @param factoryId The widget factory id.
+     * @param options The widget factory specific information.
+     *
+     * @returns Whether the widget exists or not.
+     */
+    widgetExists(factoryId: string, options?: any): boolean {
```

Comment:
I think this is fine, but there are two alternatives:
keep the `widgetExists` method, but just do this:
```typescript
const key = this.toKey({factoryId, options});
return !!this.doGetWidget(key);
```
Or, alternatively, change the `getWidget()` code slightly:
```typescript
    getWidget<T extends Widget>(factoryId: string, options?: any): MaybePromise<T> | undefined {
        const key = this.toKey({ factoryId, options });
        return this.doGetWidget<T>(key);
    }
```
By making `getWidget` not async, so that it doesn't *necessarily* return a promise, you get the same ability to check for the existence of a widget from outside the widget manager by checking !!widgetManager.getWidget(...)`